Remove Cmm.memory_chunk.Double_u

gretay-js opened the pull request:
All current targets treat `Double_u` the same as `Double`.
`Double_u` was [introduced](https://github.com/ocaml/ocaml/commit/c02c1359461e27ba91ccd8751f5855c50f9929f7) for sparc and mips, which treated it differently from `Double`.

This PR had no regressions on [precheck # 600](https://ci.inria.fr/ocaml/job/precheck/600/), except an error on `ocaml-mingw-32`  "Device or resource busy" that looks unrelated to this PR.

gretay-js committed:
Remove Cmm.memory_chunk.Double_u

All current targets treat Double_u the same as Double.

xavierleroy commented:
Right.  The `Double` / `Double_u` distinction was useful on 32-bit platforms that don't support unaligned loads of 64-bit floats.  (64-bit floats in the OCaml heap are only 4-aligned on 32-bit platforms, while 64-bit floats in bigarrays are 8-aligned.)  If we're not expecting any future port to such a platform, we can remove the distinction.  Personally I'm not expecting any future port to any 32-bit platform, but @avsm might disagree.

gretay-js committed:
Update comment for Double

gretay-js commented:
I fixed the comment attached to Double and added this PR number to it.
I should have said  in the description of this PR "remove Double and rename Double_u to Double". 

If we need to add new targets without support for unaligned memory access (for doubles or maybe other types), we can consider representing alignment requirements of operations separately from the size of data chunks.

gretay-js commented:
@avsm do you have any objection to this change?

avsm commented:
Apologies for the delayed reply; I've been meaning to check if RISC-V RV32I or RV32E (the only 32-bit architectures I remotely care about) supports unaligned loads on 64-bit floats.

A quick look at Krste's [2016  report](https://people.eecs.berkeley.edu/~krste/papers/EECS-2016-1.pdf) says that misaligned loads and stores are explicitly allowed (Chapter 3, page 21).  It's not totally clear whether this applies to the FP extensions --  @lpw25 suggests it may be implementation defined. I had a quick peer into Godbolt to see what would happen on RV32I rv32gc gcc 8.2.0:

```
struct mis {
    double f;
};
double doit(struct mis x) {
    return (x.f * 5.0);
}
```

results in 

```
doit(mis):
        lui     a5,%hi(.LC0)
        fld     fa5,%lo(.LC0)(a5)
        fmul.d  fa0,fa0,fa5
        ret
.LC0:
        .word   0
        .word   1075052544
```

Then adding some misalignment into the structure but leaving the compiler to do padding gives us:

```
struct mis {
    short i;
    double f;
};

double doit(struct mis x) {
    return (x.f * 5.0);
}
```

```
doit(mis):
        lui     a5,%hi(.LC0)
        fld     fa5,%lo(.LC0)(a5)
        addi    sp,sp,-16
        fmul.d  fa0,fa0,fa5
        addi    sp,sp,16
        jr      ra
.LC0:
        .word   0
        .word   1075052544
```

A bit more work but not unreasonable. Now forcing the struct to be packed:

```
struct mis {
    short i;
    double f;
} __attribute__((packed));

double doit(struct mis x) {
    return (x.f * 5.0);
}
```

```
doit(mis):
        addi    sp,sp,-32
        fsd     fa0,8(sp)
        lw      a2,8(sp)
        li      a4,65536
        addi    a4,a4,-1
        srli    a5,a2,16
        and     a1,a5,a4
        lw      a5,24(sp)
        lw      a3,12(sp)
        sh      a0,20(sp)
        li      a0,-65536
        and     a5,a5,a0
        or      a5,a5,a1
        and     a1,a4,a3
        and     a2,a4,a2
        slli    a1,a1,16
        and     a5,a5,a4
        or      a5,a5,a1
        slli    a1,a2,16
        lw      a2,20(sp)
        srli    a3,a3,16
        and     a6,a5,a4
        and     a2,a2,a4
        or      a2,a2,a1
        and     a4,a3,a4
        srli    a5,a5,16
        srli    a2,a2,16
        slli    a6,a6,16
        slli    a4,a4,16
        or      a1,a4,a5
        or      a0,a6,a2
        sw      a0,8(sp)
        sw      a1,12(sp)
        lui     a5,%hi(.LC0)
        fld     fa5,8(sp)
        fld     fa0,%lo(.LC0)(a5)
        addi    sp,sp,32
        fmul.d  fa0,fa5,fa0
        jr      ra
.LC0:
        .word   0
        .word   1075052544
````

(please do correct me if my increasingly creaky C is incorrect above for the purposes of the experiment conducted).

So it looks doubles being misaligned on RV32I will be slow, but not crashy. Therefore, I don't know of any 32-bit platforms in the next few years that'll need to specially handle misaligned 64-bit float loads.   We can cross the performance bridge on embedded RV32 when we come to it (I've only just received my RV64 hardware from SiFive last week).

gretay-js commented:
Thank you for looking into it.

xavierleroy commented:
Your test would be closer to what happens in OCaml with `int i` instead of `short i`, so that `f` is 4-aligned but not 8-aligned.

At any rate,  GCC generates rather pessimal code here (even with my proposed change) --- it seems to assume that `f` can have any alignment whatsoever.  

Loading a 4-aligned double-precision float can be done as two 32-bit integer loads, two 32-bit integer stores in a (8-aligned) stack temporary, and a 64-bit FP load from the stack.

The purpose of the `Double` / `Double_u` distinction was to avoid the cost of doing so when the FP number is known to be 8-aligned, i.e. comes from a bigarray and not from the OCaml heap.

At any rate, @avsm seems OK with whatever it will cost on an hypothetical RISC-V 32 port.  I'm OK too.

xavierleroy reviewed (approved):
Looks good to me.  Please add a Changes entry and we'll squash and merge.

gretay-js committed:
Update Changes

gretay-js commented:
> Looks good to me. Please add a Changes entry 

Done. Thank you!
